dft: adding initial schema for DFT data sharing

fgaray opened the pull request:
This is a WIP but good to merge now.

If this looks good @maliberty , I will followup with the custom implementation of the getters/setters, but I want to first check that I am doing the right thing with the schema/json files.

Following your comment on: https://github.com/The-OpenROAD-Project/OpenROAD/discussions/4269#discussioncomment-7581928

Thanks!

fgaray commented:
Also @maliberty Are we fixing the clang-tidy messages for auto-generated code?

maliberty commented:
> Also @maliberty Are we fixing the clang-tidy messages for auto-generated code?

Don't worry about those.

maliberty commented:
Something odd has happened here - you seem to have merged master but the diffs now show up as changes in this PR.  Did you use git merge?

fgaray commented:
> Something odd has happened here - you seem to have merged master but the diffs now show up as changes in this PR. Did you use git merge?

I am not sure what I did. For some reason, I can't push my changes to this branch but I still have them on local. I may need to do a force push to sync this. Are you ok with me doing that? I know that it may cause some issues with the review here in github.

maliberty commented:
Ideally I would avoid force pushing after a review has started.  However I think you should in this case things are messed up.

maliberty commented:
Note there was a schema update yesterday so you'll need to resolve any conflicts.

fgaray commented:
Hi @maliberty , thanks letting me force push this time. I think this is now ready for review. I am able to see the changes now.

maliberty commented on src/odb/src/codeGenerator/schema/scan/dbScanPin.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "dbScanPin",
+  "type": "dbObject",
+  "fields": [
+    {
+      "name": "bterm",
+      "type": "dbId<dbBTerm>",
+      "flags": ["private"]
+    },
+    {
+      "name": "iterm",
+      "type": "dbId<dbITerm>",
+      "flags": ["private"]
+    }
```
Should this be a std::variant ?

maliberty replied:
Yes

maliberty commented on src/odb/include/odb/db.h:
```diff
@@ -7524,6 +7528,38 @@ class dbPowerSwitch : public dbObject
+  uint getBits() const;
+};
+
+class dbScanPartition : public dbObject
+{
+ public:
+};
+
+class dbScanPin : public dbObject
+{
+ public:
+};
```
There is no way to create these objects or get/set any fields.  dbScanInst & dbScanChain  are only slightly better.  I think you need to add some custom methods here.

maliberty replied:
Ok so long as you feel you won't need to change the schema again

fgaray commented on src/odb/src/db/dbScanPin.cpp:
```diff
@@ -0,0 +1,184 @@
+// User Code Begin Includes
+namespace {
+template <class>
+inline constexpr bool always_false_v = false;
```
I think this one is a false positive. If I remove this line I will get a compile error in line 168 in this file.

fgaray commented:
Hi @maliberty , this is ready for another look.

I changed and implemented some methods in dbScanPin using std::variant, the rest of the schema is still empty as I plan to do them in a next PR. Could you please see my implementations for the serialization (>>, <<)?

Thanks!
